origin-growth/src/rules/rules.js

```diff
@@ -0,0 +1,356 @@
+           event.status === GrowthEventStatuses.Verified))
+      })
+      .forEach(event => {
+         tally[event.type] = tally.hasOwnProperty(event.type) ? tally[event.type] + 1 : 1
+      })
+    return tally
+  }
+
+  getRewards(ethAddress, events) {
+    // If this rule does not give out reward, return right away.
+    if (!this.reward) {
+      return []
+    }
+
+    const numRewards = this._numRewards(ethAddress, events)
+    const rewards = Array(numRewards).fill(this.reward)
+
+    return rewards
+  }
+}
+
+/**
+ * A rule that requires 1 event.
+ */
+class SingleEventRule extends BaseRule {
+  constructor(campaignId, levelId, config) {
+    super(campaignId, levelId, config)
+
+    const eventType = this.config.eventType
+    if (!eventType) {
+      throw new Error(`${this.str()}: missing eventType field`)
+    } else if (!GrowthEventTypes.includes(eventType)) {
+      throw new Error(`${this.str()}: unknown eventType ${eventType}`)
+    }
+    this.eventTypes = [eventType]
+  }
+
+  /**
+   * Returns number of rewards user qualifies for, taking into account the rule's limit.
+   * @param {string} ethAddress - User's account.
+   * @param {List<models.GrowthEvent>} events
+   * @returns {number}
+   * @private
+   */
+  _numRewards(ethAddress, events) {
+    const tally = this._tallyEvents(ethAddress, this.eventTypes, events)
+    return Object.keys(tally).length ? Math.min(...Object.values(tally), this.limit) : 0
+  }
+
+  /**
+   * Calculates if the rule passes.
+   * @param {string} ethAddress - User's account.
+   * @param {List<models.GrowthEvent>} events
+   * @returns {boolean}
+   */
+  evaluate(ethAddress, events) {
+    const tally = this._tallyEvents(ethAddress, this.eventTypes, events)
+    return (Object.keys(tally).length === 1)
+  }
+}
+
+/**
+ * A rule that requires N events out of a list of event types.
+ */
+class MultiEventsRule extends BaseRule {
+  constructor(campaignId, levelId, config) {
+    super(campaignId, levelId, config)
+
+    if (!this.config.eventTypes) {
+      throw new Error(`${this.str()}: missing eventTypes field`)
+    }
+    this.config.eventTypes.forEach(eventType => {
+      if (!GrowthEventTypes.includes(eventType)) {
+        throw new Error(`${this.str()}: unknown eventType ${eventType}`)
+      }
+    })
+    this.eventTypes = this.config.eventTypes
+
+    if (!this.config.numEventsRequired ||
+      !Number.isInteger(this.config.numEventsRequired) ||
+      this.config.numEventsRequired > this.eventTypes.length) {
+      throw new Error(`${this.str()}: missing or invalid numEventsRequired`)
+    }
+    this.numEventsRequired = this.config.numEventsRequired
+  }
+
+  /**
+   * Returns number of rewards user qualifies for, taking into account the rule's limit.
+   * @param {string} ethAddress - User's account.
+   * @param {List<models.GrowthEvent>} events
+   * @returns {number}
+   * @private
+   */
+  _numRewards(ethAddress, events) {
```

Comment:
I guess this is down to taste... A more functional alternative to this function:
```
_numRewards(ethAddress, events) {
  const tally = this._tallyEvents(ethAddress, this.eventTypes, events)
  if (Object.values(tally).length < this.numEventsRequired)
    return 0

 return Math.min(this.limit, ...Object.values(tally) // keep only event counts
    .sort((first, second) => second - count)  // order them by count descending
    .slice(0, this.numEventsRequired)  // keep only the first this.numEventsRequired entries
 )}  
```

Reply:
Not certain the alternative work. Any combination of events can make a tuple.
Let's say numEventsRequired=2 and we get the following tally:
{
  'A': 10,
  'B': 5,
  'C': 2,
   'D': 1
}

The propose change would return 5.

But the correct result is 8 since we can make up to 8 pairs. For example:
5 * (A, B)
2 * (A, C)
1 * (A, D)